### ISSUE TITLE
[Bug] PETS not working

### ISSUE THREAD

**Issue by sofan110** [facebookresearch/mbrl-lib#136]
## Steps to reproduce

  1. install mbrl with python3.8 & mujoco_py 2.0.2.0
  2. python -m mbrl.examples.main algorithm=pets overrides=pets_halfcheetah 
 
## Observed Results

 env_step,episode_reward,step
1000.0,-224.74164192363065,1
2000.0,-216.55716608141833,2
3000.0,-23.61229154142554,3
4000.0,-226.04264782442579,4
5000.0,299.97272326884257,5
6000.0,-424.2352836475372,6
7000.0,-605.4988140825888,7
8000.0,-276.8960448750668,8
9000.0,-570.0111469500497,9
10000.0,-510.15227529837796,10
11000.0,-521.2191905188236,11
12000.0,-380.6738015630948,12
13000.0,-401.0656166902861,13
14000.0,-342.89326195274214,14
15000.0,-387.0973047072805,15
16000.0,271.654545187927,16
17000.0,-357.9662191309233,17
18000.0,-144.4911364581224,18
19000.0,-227.65608581868534,19
20000.0,-270.1466421280269,20
21000.0,-218.2495164661332,21
22000.0,-291.59770272027646,22
23000.0,5.605493817390425,23
24000.0,-260.5804876267262,24
25000.0,-311.1006996761441,25
26000.0,-87.68273024315891,26
27000.0,-224.6058292677028,27
28000.0,-243.66672977662145,28
29000.0,-417.3611859069211,29
30000.0,-205.45597669987774,30
31000.0,-220.6631462332176,31
32000.0,-306.92107250798256,32
33000.0,-321.6192194136308,33
34000.0,156.56899647240394,34
35000.0,-373.6946869809165,35
36000.0,-297.54081355112413,36
37000.0,-403.86887923659464,37
38000.0,-394.61809157238,38
39000.0,-397.597218596027,39
40000.0,-270.5546716816992,40
41000.0,-275.0500238719418,41
42000.0,-339.1503604637613,42
43000.0,-394.371951392158,43
44000.0,-284.8456374765922,44
45000.0,-230.30455468451476,45
46000.0,-452.69669066476587,46
47000.0,-369.8052064885858,47
48000.0,-277.8216601977107,48
49000.0,83.44271984210994,49
50000.0,-165.98679718221237,50
51000.0,-286.4235189537889,51
52000.0,-420.1238034618763,52
53000.0,-348.4956325925755,53
54000.0,-262.9499726805828,54
55000.0,-82.70856034802993,55
56000.0,-283.44756999937294,56
57000.0,-296.14589401299133,57
58000.0,-310.71395667647914,58
59000.0,-92.32547170477757,59
60000.0,-343.62926472041903,60
61000.0,194.0718436837866,61
62000.0,-449.34500076620725,62
63000.0,-317.03787784175205,63
64000.0,-203.2571831873085,64
65000.0,-90.52911874178189,65
66000.0,-188.53310534801767,66
67000.0,-131.71672373665217,67
68000.0,-241.95741966590174,68
69000.0,-329.25808904770525,69
70000.0,-146.0802349071957,70
71000.0,-474.47665284478336,71
72000.0,-191.43021635327702,72

## Expected Results

  like results in #97

**Comment by luisenp** [comment]
Just to double check, do you mean MBPO? (title says PETS, but command is for MBPO)

**Comment by sofan110** [comment]
> Just to double check, do you mean MBPO? (title says PETS, but command is for MBPO)

Sorry, I made a mistake, the command is for PETS.

**Comment by jity16** [comment]
I have the same problem. PETS cannot work in halfcheetah environment with the original parameters provided by this repo. 
This is the episode reward:
![image](https://user-images.githubusercontent.com/35065354/145710757-b16f7297-cf86-4322-b725-973e86adfc75.png)

**Comment by luisenp** [comment]
Thanks a lot for reporting! We recently changed the way config files are read, so hopefully this is just a config error. I'll look into this tomorrow. Are you testing with the current version in the repo or is it the pypi version (0.1.4)?

**Comment by sofan110** [comment]
> Thanks a lot for reporting! We recently changed the way config files are read, so hopefully this is just a config error. I'll look into this tomorrow. Are you testing with the current version in the repo or is it the pypi version (0.1.4)?

I can reproduce the bug both with the repo and the PyPI version (0.1.4). The older PyPI versions are not tested.

**Comment by luisenp** [comment]
Thanks for checking pypi as well @sofan110, that's really useful.

**Comment by natolambert** [comment]
Just wanted to add that we have had some difficulties on a research fork of this that seems like a similar issue -- checking the versions and will get back to you.

**Comment by jity16** [comment]
> Quick update: version v0.1.3 works for me (testing on 5 seeds), using torch 1.10.0, cuda 10.2 and python 3.7.8. The command I'm running is below (and in fact, these hyperparms don't work for me in the latest MBRL version).
> 
> In v0.1.4 we changed some of the planning code. Our tests on ground truth dynamics were still working correctly, but maybe we inadvertently broke PETS w/o realizing. I'll keep looking at the possible cause.
> 
> In the mean time, would you mind checking if this command works in your system with MBRL v0.1.3?
> 
> ```shell
> python -m mbrl.examples.main experiment=v013_test \
>   overrides=pets_halfcheetah \
>   algorithm=pets \
>   dynamics_model.model.num_layers=4 \
>   dynamics_model.model.ensemble_size=5 \
>   dynamics_model.model.deterministic=false \
>   overrides.env=pets_halfcheetah overrides.obs_process_fn=mbrl.env.pets_halfcheetah.HalfCheetahEnv.preprocess_fn
>   overrides.learned_rewards=false overrides.no_delta_list=[0] \
>   dynamics_model.model.propagation_method=random_model \
>   +algorithm.bootstrap_permutes=false \
>   overrides.validation_ratio=0 \
>   overrides.freq_train_model=1000 \
>   overrides.patience=12 \
>   overrides.num_epochs_train_model=12 \
>   overrides.model_lr=0.00028 \
>   overrides.model_wd=0.00010 \
>   overrides.planning_horizon=30 \
>   overrides.cem_alpha=0.12 \
>   overrides.cem_population_size=400 \
>   overrides.cem_elite_ratio=0.16 \
>   overrides.cem_num_iters=5
> ```

Yes, it works in v 0.1.3! (and v 0.1.4 still fails)

**Comment by sofan110** [comment]
> > Quick update: version v0.1.3 works for me (testing on 5 seeds), using torch 1.10.0, cuda 10.2 and python 3.7.8. The command I'm running is below (and in fact, these hyperparms don't work for me in the latest MBRL version).
> > In v0.1.4 we changed some of the planning code. Our tests on ground truth dynamics were still working correctly, but maybe we inadvertently broke PETS w/o realizing. I'll keep looking at the possible cause.
> > In the mean time, would you mind checking if this command works in your system with MBRL v0.1.3?
> > ```shell
> > python -m mbrl.examples.main experiment=v013_test \
> >   overrides=pets_halfcheetah \
> >   algorithm=pets \
> >   dynamics_model.model.num_layers=4 \
> >   dynamics_model.model.ensemble_size=5 \
> >   dynamics_model.model.deterministic=false \
> >   overrides.env=pets_halfcheetah overrides.obs_process_fn=mbrl.env.pets_halfcheetah.HalfCheetahEnv.preprocess_fn
> >   overrides.learned_rewards=false overrides.no_delta_list=[0] \
> >   dynamics_model.model.propagation_method=random_model \
> >   +algorithm.bootstrap_permutes=false \
> >   overrides.validation_ratio=0 \
> >   overrides.freq_train_model=1000 \
> >   overrides.patience=12 \
> >   overrides.num_epochs_train_model=12 \
> >   overrides.model_lr=0.00028 \
> >   overrides.model_wd=0.00010 \
> >   overrides.planning_horizon=30 \
> >   overrides.cem_alpha=0.12 \
> >   overrides.cem_population_size=400 \
> >   overrides.cem_elite_ratio=0.16 \
> >   overrides.cem_num_iters=5
> > ```
> 
> Yes, it works in v 0.1.3! (and v 0.1.4 still fails)

v0,1,3 still cannot work for me, using torch 1.10.0, cuda 10.2, and python 3.8.12. The dynamics_model hyperparameters in the command connot be set in v0.1.3 and I remove them in my experiment. It is really strange for me.

**Comment by luisenp** [comment]
@sofan110, oh sorry, i should have mentioned that your config files also need to be on v0.1.3, so probably easiest way to test would be to get v0.1.3 directly from the Github stored revisions. 

Note that the config structure in v0.1.3 was different from the current one. In the new configs you have to do `dynamics_model.num_layers=4`, but in v0.1.3 you had to do `dynamics.model.num_layers=4`. That is, the previous configs had one extra nesting for dynamics model. 

Maybe this is the cause of the error you describe?

**Comment by luisenp** [comment]
Update: I took a look at the planning code and ran some quick tests, but haven't spotted any obvious differences that could cause this behavior. So, I'm currently trying to identify the commit were things broke, and hopefully that'll bring some clarity. 

@sofan110 were you able to run v0.1.3 successfully?

**Comment by Hyperion-shuo** [comment]
I have the same problem running pets cem using default config only change model ensemble to single model in v0.1.4. And i run pets successsfully on other three pets env, cartpole_continuous,  pets_pusher, pets_reacher,  gym_halfcheetah also runs well.
![image](https://user-images.githubusercontent.com/55169296/146321198-783c5c1b-5d67-48a7-b5c6-d7d599a3fd50.png)

**Comment by sofan110** [comment]
> Update: I took a look at the planning code and ran some quick tests, but haven't spotted any obvious differences that could cause this behavior. So, I'm currently trying to identify the commit were things broke, and hopefully that'll bring some clarity.
> 
> @sofan110 were you able to run v0.1.3 successfully?

Thanks for reminding me of the config files, the v0.1.3 works for me now!

**Comment by luisenp** [comment]
[This](https://github.com/facebookresearch/mbrl-lib/commit/da0c5373221b9aa3583d7f741eeb076823f40e11) seems to be the offending commit for `pets_halfcheetah`. I'll look deeper today, hopefully figure out what was the breaking change there.

**Comment by luisenp** [comment]
Pushed a [commit](https://github.com/facebookresearch/mbrl-lib/commit/c6ebdece958e0a308e3c66130f79d0bcfdde226c) that fixed the issue for me. It was caused by incorrectly using the observation pre-processing function in the `OneDTransitionRewardModel.sample()` method. @sofan110 @jity16 if you can confirm that it works for you that would be helpful! 

You can run with `python -m mbrl.examples.main algorithm=pets overrides=pets_halfcheetah`.

**Comment by jity16** [comment]
> Pushed a [commit](https://github.com/facebookresearch/mbrl-lib/commit/c6ebdece958e0a308e3c66130f79d0bcfdde226c) that fixed the issue for me. It was caused by incorrectly using the observation pre-processing function in the `OneDTransitionRewardModel.sample()` method. @sofan110 @jity16 if you can confirm that it works for you that would be helpful!
> 
> You can run with `python -m mbrl.examples.main algorithm=pets overrides=pets_halfcheetah`.

![image](https://user-images.githubusercontent.com/35065354/146745833-b67f13c1-73f0-4ce4-abc1-68f25ed576b5.png)
Yes, it works for me by changing the `OneDTransitionRewardModel.sample()` method.

**Comment by sofan110** [comment]
> Pushed a [commit](https://github.com/facebookresearch/mbrl-lib/commit/c6ebdece958e0a308e3c66130f79d0bcfdde226c) that fixed the issue for me. It was caused by incorrectly using the observation pre-processing function in the `OneDTransitionRewardModel.sample()` method. @sofan110 @jity16 if you can confirm that it works for you that would be helpful!
> 
> You can run with `python -m mbrl.examples.main algorithm=pets overrides=pets_halfcheetah`.

I also can reproduce it successfully. Thanks with sincere gratitude.
